File: app/src/main/java/com/openclassrooms/realestatemanager/database/todolist/PropertyViewModel.java
package com.openclassrooms.realestatemanager.database.todolist;


import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

import com.openclassrooms.realestatemanager.database.repositories.ImageDataRepository;
import com.openclassrooms.realestatemanager.database.repositories.PropertyDataRepository;
import com.openclassrooms.realestatemanager.database.repositories.VideoDataRepository;
import com.openclassrooms.realestatemanager.model.Image_property;
import com.openclassrooms.realestatemanager.model.Property;
import com.openclassrooms.realestatemanager.model.Video_property;

import java.util.List;
import java.util.concurrent.Executor;

public class PropertyViewModel extends ViewModel {

    // REPOSITORIES
    private final PropertyDataRepository propertyDataSource;
    private final VideoDataRepository videoDataSource;
    private final ImageDataRepository imageDataSource;


    private final Executor executor;

    // DATA
    @Nullable
    private LiveData<List<Property>> currentProperties;

    public PropertyViewModel(PropertyDataRepository propertyDataSource, VideoDataRepository videoDataSource, ImageDataRepository imageDataSource, Executor executor) {
        this.propertyDataSource = propertyDataSource;
        this.videoDataSource = videoDataSource;
        this.imageDataSource = imageDataSource;
        this.executor = executor;
    }

    public void init() {
        if (this.currentProperties != null) {
            return;
        }
        currentProperties = propertyDataSource.getProperties();
    }


    public LiveData<List<Property>> getProperties() {
        return propertyDataSource.getProperties() ;
    }

    public LiveData<List<Video_property>> getVideos() {
        return videoDataSource.getVideos() ;

    }

    public void createVideo(Video_property video) {
        executor.execute(() -> {
            videoDataSource.createVideo(video);
        });
    }

    public void deleteVideo(int videoId) {
        executor.execute(() -> {
            videoDataSource.deleteVideo(videoId);
        });
    }



    public LiveData<List<Image_property>> getImages() {
        return imageDataSource.getImages() ;

    }

    public void createImage(Image_property image) {
        executor.execute(() -> {
            imageDataSource.createImage(image);
        });
    }

    public void deleteImage(int imageId) {
        executor.execute(() -> {
            imageDataSource.deleteImage(imageId);
        });
    }







   public LiveData<List<Property>> findCorrectProperties(String type,int priceMin, int surfaceMin, int pieceMin, int priceMax, int surfaceMax, int pieceMax, String descript, String ville, String address, String proximity, String statu, String startDate, String sellingDate, String agent, String isDollar, int photoMin, int photoMax, int videoMin, int videoMax) {
        return propertyDataSource.findCorrectProperties( type, priceMin,  surfaceMin, pieceMin,  priceMax, surfaceMax,  pieceMax, descript,  ville,  address,  proximity,  statu,  startDate,  sellingDate, agent,  isDollar,  photoMin,  photoMax,  videoMin, videoMax) ;
    }


    public void createProperty(Property property) {
        executor.execute(() -> {
            propertyDataSource.createProperty(property);
        });
    }



}

File: app/src/main/java/com/openclassrooms/realestatemanager/database/todolist/PropertyViewModel.kt
package com.openclassrooms.realestatemanager.database.todolist


import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel

import com.openclassrooms.realestatemanager.database.repositories.ImageDataRepository
import com.openclassrooms.realestatemanager.database.repositories.PropertyDataRepository
import com.openclassrooms.realestatemanager.database.repositories.VideoDataRepository
import com.openclassrooms.realestatemanager.model.Image_property
import com.openclassrooms.realestatemanager.model.Property
import com.openclassrooms.realestatemanager.model.Video_property
import java.util.concurrent.Executor

class PropertyViewModel(// REPOSITORIES
        private val propertyDataSource: PropertyDataRepository, private val videoDataSource: VideoDataRepository, private val imageDataSource: ImageDataRepository, private val executor: Executor) : ViewModel() {

    // DATA
    private var currentProperties: LiveData<List<Property>>? = null


    val properties: LiveData<List<Property>>
        get() = propertyDataSource.properties

    val videos: LiveData<List<Video_property>>
        get() = videoDataSource.videos


    val images: LiveData<List<Image_property>>
        get() = imageDataSource.images

    fun init() {
        if (this.currentProperties != null) {
            return
        }
        currentProperties = propertyDataSource.properties
    }

    fun createVideo(video: Video_property) {
        executor.execute { videoDataSource.createVideo(video) }
    }

    fun deleteVideo(videoId: Int) {
        executor.execute { videoDataSource.deleteVideo(videoId) }
    }

    fun createImage(image: Image_property) {
        executor.execute { imageDataSource.createImage(image) }
    }

    fun deleteImage(imageId: Int) {
        executor.execute { imageDataSource.deleteImage(imageId) }
    }


    fun findCorrectProperties(type: String, priceMin: Int, surfaceMin: Int, pieceMin: Int, priceMax: Int, surfaceMax: Int, pieceMax: Int, descript: String, ville: String, address: String, proximity: String, statu: String, startDate: String, sellingDate: String, agent: String, isDollar: String, photoMin: Int, photoMax: Int, videoMin: Int, videoMax: Int): LiveData<List<Property>> {
        return propertyDataSource.findCorrectProperties(type, priceMin, surfaceMin, pieceMin, priceMax, surfaceMax, pieceMax, descript, ville, address, proximity, statu, startDate, sellingDate, agent, isDollar, photoMin, photoMax, videoMin, videoMax)
    }


    fun createProperty(property: Property) {
        executor.execute { propertyDataSource.createProperty(property) }
    }


}